include/audio_vector.h

```diff
@@ -0,0 +1,192 @@
+		  data(_data, _data + _num_frames) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		channel->AddSamples_m8s(num_frames, data.data());
+	}
+private:
+	std::vector<int8_t> data = {};
+};
+
+class AudioVectorS8 final : public AudioVector
+{
+public:
+	AudioVectorS8(int _num_frames, const uint8_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + (_num_frames * 2)) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		channel->AddSamples_s8(num_frames, data.data());
+	}
+private:
+	std::vector<uint8_t> data = {};
+};
+
+class AudioVectorS8S final : public AudioVector
+{
+public:
+	AudioVectorS8S(int _num_frames, const int8_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + (_num_frames * 2)) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		channel->AddSamples_s8s(num_frames, data.data());
+	}
+private:
+	std::vector<int8_t> data = {};
+};
+
+class AudioVectorS16 final : public AudioVector
+{
+public:
+	AudioVectorS16(int _num_frames, const int16_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + (_num_frames * 2)) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		#if defined(WORDS_BIGENDIAN)
+		channel->AddSamples_s16_nonnative(num_frames, data.data());
+		#else
+		channel->AddSamples_s16(num_frames, data.data());
+		#endif
+	}
+private:
+	std::vector<int16_t> data = {};
+};
+
+class AudioVectorS16U final : public AudioVector
+{
+public:
+	AudioVectorS16U(int _num_frames, const uint16_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + (_num_frames * 2)) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		#if defined(WORDS_BIGENDIAN)
+		channel->AddSamples_s16u_nonnative(num_frames, data.data());
+		#else
+		channel->AddSamples_s16u(num_frames, data.data());
+		#endif
+	}
+private:
+	std::vector<uint16_t> data = {};
+};
+
+class AudioVectorM16 final : public AudioVector
+{
+public:
+	AudioVectorM16(int _num_frames, const int16_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + _num_frames) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		#if defined(WORDS_BIGENDIAN)
+		channel->AddSamples_m16_nonnative(num_frames, data.data());
+		#else
+		channel->AddSamples_m16(num_frames, data.data());
+		#endif
+	}
+private:
+	std::vector<int16_t> data = {};
+};
+
+class AudioVectorM16U final : public AudioVector
+{
+public:
+	AudioVectorM16U(int _num_frames, const uint16_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + _num_frames) {}
+
+	void AddSamples(MixerChannel* channel) final
+	{
+		#if defined(WORDS_BIGENDIAN)
+		channel->AddSamples_m16u_nonnative(num_frames, data.data());
+		#else
+		channel->AddSamples_m16u(num_frames, data.data());
+		#endif
+	}
+private:
+	std::vector<uint16_t> data = {};
+};
+
+class AudioVectorStretched final : public AudioVector {
+public:
+	AudioVectorStretched(int _num_frames, const int16_t* _data)
+		: AudioVector(_num_frames),
+		  data(_data, _data + _num_frames) {}
+
```

Comment:
I don't know what you mean by that.  This is a special class that will only ever be used by SoundBlaster DAC.  It calls `channel->AddStreched()` whereas everything else calls a templated version of `channel->AddSamples()`.  I'm not doing anything special with endianess here.

Unlikely that any of these `AudioVector` variants will be used outside of SoundBlaster as it's inefficient with heap allocations and virtual functions (though could be done if the need arises).  Every other audio device enqueues frames directly without this wrapping class.

Reply:
> I don't know what you mean by that. 

It's a header in include/, and there wasn't any SB-only comments or names in the class.  So to me, it appears to be for general use for any audio code. 

So I wanted to mention that the assumption about always byte swapping the 16bit types won't work for host-generated 16-bit values. 


> This is a special class that will only ever be used by SoundBlaster DAC.

Ok.. then this makes sense. 
If it's SB only, then suggest adding comments to that effect and maybe move the header alone side sblaster.cpp, because it's not a general data structure.

Reply:
Stretched is the only one that is really SoundBlaster specific and even that is generic if some new audio device wants to stretch samples, provided it uses the same audio format.  Probably that won't ever happen though.

The rest are very generic.  The 16 bit types are native endianess which seems like a sane default.  It could be extended to have little/big endian types but I'm not adding that until we have a use case.

Reply:
> The rest are very generic. The 16 bit types are native endianess which seems like a sane default.

The 16 bit handling treats the incoming value as little endian even on big endian hosts (where the incoming value is byte swapped). 

If the incoming 16bit was treated as native, the byte swapping logic wouldn't exist.

This is all correct for sound blaster and GUS because DOS is always little endian and the games setup 16bit samples for DMA as little endian.  

But when it comes to the synth code like OPL etc.. that code is generating 16bit in samples in native endian, so that's why I'm saying this logic isn't general and shouldn't be used for the 16bit synths.

Reply:
> but I'm not adding that until we have a use case.

Totally agree. I'm only suggesting strengthening (or warning) about the 16bit handling as a comment.
 - can be used for DOS DMA (alway little endian)
 - can be used for CDDA raw samples from bin/cue (always little endian)
 - cannot be used for 16bit synths (host native endian)
 - cannot be used for compressed CDDA tracks because those codecs give back 16bit host native endian samples)

Reply:
My suggestion is to leave this alone, the world is little endian and that will hardly change. BE is only used in obscure old hardware that Staging barely runs on (and we don't support BE officially, it's community best-effort).

I'd rather just change the SDL sample format from `AUDIO_S16SYS` to `AUDIO_S16LSB` for now to "pin down" the endianness, but we'll move to feeding SDL floats soon anyway (so `AUDIO_F32`).

Reply:
> My suggestion is to leave this alone,

Let me resolve this and I'll make a tiny suggestion to the code-comment; I did a poor job of being concise :-)

> Then the integer to float conversions will always be the audio devices' responsibility 

Yup - the sooner we eliminate all the int-type and endian handling from the mixer the better :rocket: 

> but we'll move to feeding SDL floats soon anyway (so AUDIO_F32).

`AudioFrame` all day every day :sunglasses: 
When I saw all that conversion code, it will be really nice to eliminate all that!